File: datastructures/linkedlist/java/algorithms/LinkedListIntersection.java
package datastructures.linkedlist.java.algorithms;

import java.util.HashSet;
import java.util.Set;

public class LinkedListIntersection {

    public static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        Set<ListNode> visited = new HashSet<>();

        ListNode current = headA;
        while (current != null) {
            visited.add(current);
            current = current.next;
        }

        current = headB;
        while (current != null) {
            if (visited.contains(current)) {
                return current;
            }
            current = current.next;
        }

        return null;
    }

    public static ListNode getIntersectionNodeOptimal(ListNode headA, ListNode headB) {
        ListNode pA = headA;
        ListNode pB = headB;

        while (pA != pB) {
            pA = pA == null ? headB : pA.next;
            pB = pB == null ? headA : pB.next;
        }

        return pA;
    }

    public static void main(String[] args) {
        ListNode headA = new ListNode(1);
        ListNode nodeA1 = new ListNode(2);
        ListNode nodeA2 = new ListNode(3);
        ListNode nodeA3 = new ListNode(4);

        headA.next = nodeA1;
        nodeA1.next = nodeA2;
        nodeA2.next = nodeA3;

        ListNode headB = new ListNode(5);
        ListNode nodeB1 = new ListNode(6);
        ListNode nodeB2 = new ListNode(7);

        headB.next = nodeB1;
        nodeB1.next = nodeB2;

        ListNode node1 = new ListNode(8);
        ListNode node2 = new ListNode(9);
        ListNode node3 = new ListNode(10);

        nodeA3.next = node1;
        nodeB2.next = node1;
        node1.next = node2;
        node2.next = node3;

        System.out.println(getIntersectionNodeOptimal(headA, headB).val);
    }

}

File: datastructures/linkedlist/python/algorithms/twopointer/linked-list-intersection.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def get_intersection_node(headA, headB):
    pA = headA
    pB = headB

    while pA != pB:
        pA = headB if pA is None else pA.next
        pB = headA if pB is None else pB.next

    return pA


if __name__ == "__main__":
    head_A = ListNode(1)
    node_A1 = ListNode(2)
    node_A2 = ListNode(3)
    node_A3 = ListNode(4)

    head_A.next = node_A1
    node_A1.next = node_A2
    node_A2.next = node_A3

    head_B = ListNode(5)
    node_B1 = ListNode(6)
    node_B2 = ListNode(7)

    head_B.next = node_B1
    node_B1.next = node_B2

    tail_1 = ListNode(8)
    tail_2 = ListNode(9)
    tail_3 = ListNode(10)

    node_A3.next = tail_1
    node_B2.next = tail_1
    tail_1.next = tail_2
    tail_2.next = tail_3

    print(get_intersection_node(head_A, head_B).val)
